scripts/install/sonarrv3.sh

```diff
@@ -0,0 +1,179 @@
+#!/bin/bash
+# Sonarr v3 installer
+# Flying sauasges for swizzin 2020
+
+if [[ -f /tmp/.install.lock ]]; then
+  export log="/root/logs/install.log"
+else
+  log="/root/logs/swizzin.log"
+fi
+
+#shellcheck source=sources/functions/ask
+. /etc/swizzin/sources/functions/ask
+
+#Handles existing v2 instances
+_sonarrv2_flow(){
+    v2present=false
+    if [[ -f /install/.sonarr.lock ]]; then
+        v2present=true
+    fi
+    if dpkg -l | grep nzbdrone > /dev/null 2>&1 ; then
+        v2present=true
+    fi
+
+    if [[ $v2present == "true" ]]; then
+        echo "Sonarr v2 is detected."
+        echo "Continuing will migrate your current v2 installation." | tee -a $log
+        echo "You can read more about the migration at https://docs.swizzin.ltd/applications/sonarrv3#migrating-from-v2"
+        echo "An additional copy of the backup will be made into /root/sonarrv2.bak/" | tee -a $log
+        if ! ask "Do you want to continue?" N; then
+            exit 0
+        fi
+
+        if ask "Would you like to trigger a Sonarr-side backup?" Y; then
+            echo "Backing up Sonarr v2" | tee -a $log
+            if [[ -f /install/.nginx ]]; then
+                address="http://127.0.0.1:8989/sonarr/api"
+            else
+                address="http://127.0.0.1:8989/api"
+            fi
+
+            [[ -z $sonarrv2owner ]] && sonarrv2owner=$(cut -d: -f1 < /root/.master.info)
+            if [[ ! -d /home/"${sonarrv2owner}"/.config/NzbDrone ]]; then
+                echo "No Sonarr config folder found for $sonarrv2owner. Exiting" | tee -a $log
+                exit 1
+            fi
+
+            apikey=$(awk -F '[<>]' '/ApiKey/{print $3}' /home/"${sonarrv2owner}"/.config/NzbDrone/config.xml)
+            echo "apikey = $apikey" >> $log
+
+            #This starts a backup on the current Sonarr instance. The logic below waits until the query returns as "completed"
+            response=$(curl -sd '{name: "backup"}' -H "Content-Type: application/json" -X POST ${address}/command?apikey="${apikey}")
+            echo "$response" >> $log
+            id=$(echo "$response" | jq '.id' )
+            echo "id=$id" >> $log
+
+            if [[ -z $id ]]; then
+                echo "Failure triggering backup (see logs)." | tee -a $log
+                echo "We cannot trigger Sonarr to dump a current backup, but the current files and previous weekly backups can still be copied" | tee -a $log
+                if ! ask "Continue without triggering internal Sonarr backup?" N; then
+                    exit 1
+                fi
+            else
+                echo "Sonarr backup Job ID = $id, waiting to finish" >> $log
+
+                status=""
+                while [[ $status != "\"completed\"" ]]; do
```

Comment:
If the backup fails this will create an infinite loop.

Does sonarr assign a pending status for tasks in progress?

Reply:
It sets it as "running" or something while it is not finished. When it is, the API will return that. If it creates an infinite loop, the user will know that it's just the backup that failed and it won't make a huge issue with the installation. Could add some timeout or something if you'd prefer. How about 20 seconds?

Reply:
Added 20s counter, should do it

Reply:
I don't know if 20 seconds would be enough for some super huge libraries out there. Isn't it tarring up zillions of tiny files?

Can we not check to see if the status is `running` and then once it's no longer running we catch the response and handle it?

Reply:
Oh I see what you mean, got it. I'll test it out tomorrow. There could be some other status responses other than `completed` you are right.

Reply:
Yeah, sorry, that's what I was trying to convey. If the task returns a status other than completed, but it finished then we should handle that a bit more gracefully than purgatory :smile:

Reply:
Tested and resolved